src/parse_impl.rs

```diff
@@ -0,0 +1,167 @@
+        tokens,
+        |tt| matches!(tt, TokenTree::Punct(punct) if punct.as_char() == ';'),
+        true,
+    );
+
+    let tk_semicolon = match tokens.next() {
+        Some(TokenTree::Punct(punct)) if punct.as_char() == ';' => punct,
+        _ => panic!("cannot parse constant"),
+    };
+
+    Constant {
+        attributes,
+        vis_marker,
+        tk_const,
+        name,
+        tk_colon,
+        ty: TyExpr { tokens: ty_tokens },
+        tk_equals,
+        initializer: ValueExpr {
+            tokens: value_tokens,
+        },
+        tk_semicolon,
+    }
+}
+
+pub(crate) fn consume_assoc_ty(
+    tokens: &mut TokenIter,
+    attributes: Vec<Attribute>,
+    vis_marker: Option<VisMarker>,
+) -> TyDefinition {
+    let tk_type = match tokens.next() {
+        Some(TokenTree::Ident(ident)) if ident == "type" => ident,
+        _ => panic!("cannot parse associated type"),
+    };
+
+    let name = match tokens.next() {
+        Some(TokenTree::Ident(ident)) => ident,
+        _ => panic!("cannot parse associated type"),
+    };
+
+    let tk_equals = match tokens.next() {
+        Some(TokenTree::Punct(punct)) if punct.as_char() == '=' => punct,
+        _ => panic!("cannot parse associated type"),
+    };
+
+    let ty_tokens = consume_stuff_until(
+        tokens,
+        |tt| matches!(tt, TokenTree::Punct(punct) if punct.as_char() == ';'),
+        true,
+    );
+
+    let tk_semicolon = match tokens.next() {
+        Some(TokenTree::Punct(punct)) if punct.as_char() == ';' => punct,
+        _ => panic!("cannot parse associated type"),
+    };
+
+    TyDefinition {
+        attributes,
+        vis_marker,
+        tk_type,
+        name,
+        tk_equals,
+        initializer_ty: TyExpr { tokens: ty_tokens },
+        tk_semicolon,
+    }
+}
+
+pub(crate) fn consume_fn_const_or_type(
+    tokens: &mut TokenIter,
+    attributes: Vec<Attribute>,
+    vis_marker: Option<VisMarker>,
+    context: &str, // for panic
+) -> ImplMember {
+    if let Some(TokenTree::Ident(ident)) = tokens.peek() {
+        match ident.to_string().as_str() {
+            "type" => {
+                let assoc_ty = consume_assoc_ty(tokens, attributes, vis_marker);
+                ImplMember::AssocTy(assoc_ty)
+            }
+            "default" | "const" | "async" | "unsafe" | "extern" | "fn" => {
+                if let Some(method) = try_consume_fn(tokens, attributes.clone(), vis_marker.clone())
+                {
+                    ImplMember::Method(method)
+                } else {
+                    let constant = consume_constant(tokens, attributes, vis_marker);
+                    ImplMember::Constant(constant)
+                }
+            }
+            _ => panic!("unsupported {} item `{}`", context, ident),
+        }
+    } else {
+        panic!("unsupported {} element: {:?}", context, tokens.peek())
+    }
+}
+
+pub(crate) fn parse_impl_members(token_group: Group) -> (Vec<ImplMember>, GroupSpan) {
```

Comment:
This misses inner attributes, eg:

```rust
impl Foobar {
    #![hello]
}
```

Reply:
Implemented.

Also renamed `tk_hashbang` to `tk_hash`, which is more correct.